src_c/simd_blitters_avx2.c

```diff
@@ -12,6 +12,206 @@
+       --------------------------------------------
+       dstRGB = (srcRGB * srcA) + (dstRGB * (1-srcA))
+         dstA = srcA + (dstA * (1-srcA))
+
+       We use something slightly different to simulate
+       SDL1, as follows:
+       dstRGB = (((dstRGB << 8) + (srcRGB - dstRGB) * srcA + srcRGB) >> 8)
+         dstA = srcA + dstA - ((srcA * dstA) / 255);
+    */
+
+    /* Alpha blend procedure in this blitter:
+       --------------------------------------
+        srcRGBA = 16 bit interspersed source pixels
+            i.e. [0][A][0][R][0][G][0][B]
+        srcA = 16 bit interspersed alpha channel
+            i.e. [0][A][0][A][0][A][0][A]
+        As you can see, each pixel is moved out into 64 bits of register for
+        processing. This blitter loads in 8 pixels at once, but processes
+        4 at a time (4x64=256 bits).
+
+        Order of operations:
+            temp1 = srcRGBA - dstRGBA
+            temp1 *= srcA
+            dstRGBA <<= 8
+            dstRGBA += temp1
+            dstRGBA += srcRGBA
+            dstRGBA >>= 8
+     */
+
+    while (height--) {
+        if (pre_8_width > 0) {
+            /* ==== load 1-7 pixels into AVX registers ==== */
+            mm256_src = _mm256_maskload_epi32(srcp, mm256_mask);
+            mm256_dst = _mm256_maskload_epi32(dstp, mm256_mask);
+
+            /* ==== shuffle pixels out into two registers each, src
+             * and dst set up for 16 bit math, like 0A0R0G0B ==== */
+            mm256_srcA = _mm256_shuffle_epi8(mm256_src, mm256_shuff_mask_A);
+            mm256_srcB = _mm256_shuffle_epi8(mm256_src, mm256_shuff_mask_B);
+
+            mm256_dstA = _mm256_shuffle_epi8(mm256_dst, mm256_shuff_mask_A);
+            mm256_dstB = _mm256_shuffle_epi8(mm256_dst, mm256_shuff_mask_B);
+
+            /* ==== alpha blend opaque dst on A pixels ==== */
+            mm256_srcAlpha = _mm256_shuffle_epi8(mm256_srcA, shuffle_out);
+
+            temp = _mm256_sub_epi16(mm256_srcA, mm256_dstA);
+            temp = _mm256_mullo_epi16(temp, mm256_srcAlpha);
+
+            mm256_dstA = _mm256_slli_epi16(mm256_dstA, 8);
+            mm256_dstA = _mm256_add_epi16(mm256_dstA, temp);
+            mm256_dstA = _mm256_add_epi16(mm256_dstA, mm256_srcA);
+            mm256_dstA = _mm256_srli_epi16(mm256_dstA, 8);
+
+            /* ==== alpha blend opaque dst on B pixels ==== */
+            mm256_srcAlpha = _mm256_shuffle_epi8(mm256_srcB, shuffle_out);
+
+            temp = _mm256_sub_epi16(mm256_srcB, mm256_dstB);
+            temp = _mm256_mullo_epi16(temp, mm256_srcAlpha);
+
+            mm256_dstB = _mm256_slli_epi16(mm256_dstB, 8);
+            mm256_dstB = _mm256_add_epi16(mm256_dstB, temp);
+            mm256_dstB = _mm256_add_epi16(mm256_dstB, mm256_srcB);
+            mm256_dstB = _mm256_srli_epi16(mm256_dstB, 8);
+
+            /* ==== recombine A and B pixels and store ==== */
+            mm256_dst = _mm256_packus_epi16(mm256_dstA, mm256_dstB);
```

Comment:
Is this chunk now identical to the middle of the 8 pixel width chunk below?

Perhaps we can reduce the duplication somewhat with an inline function call, or a macro - assuming it doesn't affect performance.

Would make it clearer that only the loading and storing changes between the two widths of blend.

The masked load and store seem super useful!

Reply:
I'm open to doing that.

I was thinking about macros with these functions-- there could be a possibility where you have something like

Not even close to the right syntax, but an illustration:
```
MACRO build_avx2_blitter (name)
{arbitrary setup code}
{stride switch code}
{full force code}
```

Where the boilerplate of parsing the blitinfo and iterating is handled by the macro to reduce code duplication.

Reply:
I have high hopes for macros, but for this PR I'd like to just move forward as is. Macro+masked-load and store could be brought to the blitters in another PR.